File: 2015/Day_22/Ruby/part_2.rb
# Spells
MAGIC_MISSILE = {"cost"=> 53, "damage"=> 4}
DRAIN = {"cost"=> 73, "damage"=> 2, "heal"=> 2}
SHIELD = {"cost"=> 113, "armor"=> 7, "effect_length"=> 6, "timer"=> 0}
POISON = {"cost"=> 173, "damage"=> 3, "effect_length"=> 6, "timer"=> 0}
RECHARGE = {"cost"=> 229, "mana_refill"=> 101, "effect_length"=> 5, "timer"=> 0}

def decide_spell
  spell_list = %w[shield drain recharge poison magic_missile]
  while true
    cast_this_spell = spell_list.sample
    case cast_this_spell
    when "shield"
      if SHIELD["timer"] == 0 and WIZARD["mana_points"] >= SHIELD["cost"]
        return "shield"
      end
    when "drain"
        if WIZARD["mana_points"] >= DRAIN["cost"]
          return "drain"
        end
    when "recharge"
      if WIZARD["mana_points"] >= RECHARGE["cost"] and RECHARGE['timer'] == 0
        return "recharge"
      end
    when "poison"
      if POISON["timer"] == 0 and WIZARD["mana_points"] >= POISON["cost"]
        return "poison"
      end
    when "magic_missile"
      if WIZARD["mana_points"] >= MAGIC_MISSILE["cost"]
        return "magic_missile"
      end
    else
      puts "You should never get here."
    end
  end
end

def run_timer_spells
  #puts "running timer spells"
  if SHIELD["timer"] > 0
    SHIELD["timer"] -= 1
    WIZARD["armor"] -= SHIELD["armor"] if SHIELD["timer"] == 0
  end
  if POISON["timer"] > 0
    BOSS["hit_points"] -= POISON["damage"]
    POISON["timer"] -= 1
  end
  if RECHARGE["timer"] > 0
    WIZARD["mana_points"] += RECHARGE["mana_refill"]
    RECHARGE["timer"] -= 1
  end
end

def cast_spell(spell_name)
  case spell_name
  when "drain"
    WIZARD["mana_points"] -= DRAIN["cost"]
    WIZARD["hit_points"] += DRAIN["heal"]
    BOSS["hit_points"] -= DRAIN["damage"]
    return DRAIN["cost"]
  when "magic_missile"
    WIZARD["mana_points"] -= MAGIC_MISSILE["cost"]
    BOSS["hit_points"] -= MAGIC_MISSILE["damage"]
    return MAGIC_MISSILE["cost"]
  when "poison"
    WIZARD["mana_points"] -= POISON["cost"]
    POISON["timer"] = POISON["effect_length"]
    return POISON["cost"]
  when "recharge"
    WIZARD["mana_points"] -= RECHARGE["cost"]
    RECHARGE["timer"] = RECHARGE["effect_length"]
    return RECHARGE["cost"]
  when "shield"
    WIZARD["mana_points"] -= SHIELD["cost"]
    WIZARD["armor"] += SHIELD["armor"]
    SHIELD["timer"] = SHIELD["effect_length"]
    return SHIELD["cost"]
  else
    puts "You should never get here."
  end
end

def battle_sim
  mana_spent = 0
  while true
    # wizard turn
    #puts "wizard turn"
    # puts WIZARD
    WIZARD["hit_points"] -= 1
    return [false, mana_spent] if WIZARD["hit_points"] <= 0
    run_timer_spells
    return [false, mana_spent] if WIZARD["mana_points"] <= 53 or WIZARD["hit_points"] <= 0
    mana_spent += cast_spell(decide_spell)
    # puts mana_spent
    # boss turn
    #puts "boss turn"
    run_timer_spells
    return [true, mana_spent] if BOSS["hit_points"] <= 0
    WIZARD['hit_points'] = WIZARD['hit_points'] - (BOSS['damage'] - WIZARD['armor'])
    return [false, mana_spent] if WIZARD["hit_points"] <= 0
  end
end

mana_spent_to_win = 10000000000000000000
(1..100000).each do |trial|
  # puts trial
  WIZARD = {"hit_points"=> 50, "mana_points"=> 500, "armor"=> 0}
  BOSS = {"hit_points"=> 55, "damage"=> 8}
  SHIELD["timer"] = 0
  POISON["timer"] = 0
  RECHARGE["timer"] = 0
  did_player_win, mana_spent_this_time = battle_sim
  if did_player_win
    puts "player wins with #{mana_spent_this_time}"
    mana_spent_to_win = [mana_spent_to_win, mana_spent_this_time].min
  else
    puts "Player lost with #{mana_spent_this_time}"
  end
end
puts "Player spent #{mana_spent_to_win} mana to win."

File: 2015/Day_22/Python/part_2.py
from random import choice

# Spells
magic_missile = {"cost": 53, "damage": 4}
drain = {"cost": 73, "damage": 2, "heal": 2}
shield = {"cost": 113, "armor": 7, "effect_length": 6, "timer": 0}
poison = {"cost": 173, "damage": 3, "effect_length": 6, "timer": 0}
recharge = {"cost": 229, "mana_refill": 101, "effect_length": 5, "timer": 0}


def decide_spell():
    """Figure out which spell to cast."""
    spell_list = ["shield", "drain", "recharge", "poison", "magic_missile"]
    while True:
        cast_this_spell = choice(spell_list)
        if cast_this_spell == "shield":
            if shield["timer"] == 0 and wizard["mana_points"] >= shield["cost"]:
                print("chose shield")
                return "shield"
        elif cast_this_spell == "drain":
            if wizard["mana_points"] >= drain["cost"]:
                print("chose drain")
                return "drain"
        elif cast_this_spell == "recharge":
            if wizard["mana_points"] >= recharge["cost"] and recharge['timer'] == 0:
                print("chose recharge")
                return "recharge"
        elif cast_this_spell == "poison":
            if poison["timer"] == 0 and wizard["mana_points"] >= poison["cost"]:
                print("chose poison")
                return "poison"
        elif cast_this_spell == "magic_missile":
            if wizard["mana_points"] >= magic_missile["cost"]:
                print("chose magic missile")
                return "magic_missile"


def run_timer_spells():
    if shield["timer"] > 0:
        print("Shield active")
        shield["timer"] -= 1
        print(f'Shield timer drops to {shield["timer"]}')
        if shield["timer"] == 0:
            print("Shield has expired, lowering wizard armor")
            wizard["armor"] -= shield["armor"]
    if poison["timer"] > 0:
        print("Poison active")
        boss["hit_points"] -= poison["damage"]
        print(f"Poison attacks boss for {poison['damage']}")
        poison["timer"] -= 1
        print(f'Poison timer drops to {poison["timer"]}')
    if recharge["timer"] > 0:
        print("Recharge active")
        wizard["mana_points"] += recharge["mana_refill"]
        print(f"Mana refilled by +{recharge['mana_refill']}")
        recharge["timer"] -= 1
        print(f"Recharge drops to {recharge['timer']}")


def cast_spell(spell_name: str):
    if spell_name == "drain":
        wizard["mana_points"] -= drain["cost"]
        wizard["hit_points"] += drain["heal"]
        boss["hit_points"] -= drain["damage"]
        return drain["cost"]
    elif spell_name == "magic_missile":
        wizard["mana_points"] -= magic_missile["cost"]
        boss["hit_points"] -= magic_missile["damage"]
        return magic_missile["cost"]
    elif spell_name == "poison":
        wizard["mana_points"] -= poison["cost"]
        poison["timer"] = poison["effect_length"]
        return poison["cost"]
    elif spell_name == "recharge":
        wizard["mana_points"] -= recharge["cost"]
        recharge["timer"] = recharge["effect_length"]
        return recharge["cost"]
    elif spell_name == "shield":
        wizard["mana_points"] -= shield["cost"]
        wizard["armor"] += shield["armor"]
        shield["timer"] = shield["effect_length"]
        return shield["cost"]


def battle_sim():
    """Simulate a battle and return true if the player wins."""
    mana_spent = 0
    while True:
        print("--player turn--")
        print(f"Wizard: HP: {wizard['hit_points']}, Mana: {wizard['mana_points']}; Boss: {boss['hit_points']}")
        # hard mode - lost 1 HP before anything else happens
        wizard["hit_points"] -= 1
        if wizard["hit_points"] == 0:
            return False, mana_spent
        run_timer_spells()
        if wizard["mana_points"] <= 53 or wizard["hit_points"] <= 0:
            return False, mana_spent
        mana_spent += cast_spell(decide_spell())
        # boss turn
        print("--boss turn--")
        print(f"Wizard: HP: {wizard['hit_points']}, Mana: {wizard['mana_points']}; Boss: {boss['hit_points']}")
        run_timer_spells()
        print("Boss attacks 8")
        if boss["hit_points"] <= 0:
            # boss is dead
            return True, mana_spent
        wizard['hit_points'] = wizard['hit_points'] - (boss['damage'] - wizard['armor'])
        if wizard["hit_points"] <= 0:
            # wizard is dead
            return False, mana_spent


if __name__ == "__main__":
    mana_spent_to_win = 10000000000000000000
    for trial in range(100000):
        print(f"{trial=}")
        # set/reset wizard adn boss stats
        wizard = {"hit_points": 50, "mana_points": 500, "armor": 0}
        boss = {"hit_points": 55, "damage": 8}
        # set/reset spell timers
        shield["timer"] = 0
        poison["timer"] = 0
        recharge["timer"] = 0
        did_player_win, mana_spent_this_time = battle_sim()
        if did_player_win:
            print(f"Player won spent {mana_spent_this_time} mana")
            if mana_spent_this_time < mana_spent_to_win:
                mana_spent_to_win = mana_spent_this_time
        else:
            print(f"Player lost, spent {mana_spent_this_time} mana")
    print(f"To win required {mana_spent_to_win} mana.")
